pkg/syncer/cnsoperator/controller/triggercsifullsync/triggercsifullsync_controller.go

```diff
@@ -0,0 +1,342 @@
+		msg := fmt.Sprintf("Only %q should be used to trigger full sync and not %q",
+			common.TriggerCsiFullSyncCRName, instance.Name)
+		log.Error(msg)
+		recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+		return reconcile.Result{}, nil
+	}
+
+	// Ignore any updates on TriggerCsiFullSync instance with TriggerSyncID set to 0
+	// and TriggerSyncID same as LastTriggerSyncID
+	if instance.Spec.TriggerSyncID == 0 || instance.Spec.TriggerSyncID == instance.Status.LastTriggerSyncID {
+		return reconcile.Result{}, nil
+	}
+
+	// If TriggerSyncID is not one greater than LastTriggerSyncID, raise an event that
+	// the trigger sync will be ignored
+	if instance.Spec.TriggerSyncID != instance.Status.LastTriggerSyncID+1 {
+		msg := fmt.Sprintf("TriggerSyncID: %d is invalid. TriggerSyncID should be one greater than LastTriggerSyncID.",
+			instance.Spec.TriggerSyncID)
+		log.Error(msg)
+		recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+		return reconcile.Result{}, nil
+	}
+
+	// If the TriggerCsiFullSync instance is already in progress, update
+	// LastTriggerSyncID and raise an event that full sync is already in progress.
+	if instance.Status.InProgress && instance.Spec.TriggerSyncID == instance.Status.LastTriggerSyncID+1 {
+		// LastTriggerSyncID saves the last TriggerSyncID attempted by the
+		// user regardless of success or failure.
+		instance.Status.LastTriggerSyncID = instance.Spec.TriggerSyncID
+		err = updateTriggerCsiFullSync(ctx, r.client, instance)
+		if err != nil {
+			recordEvent(ctx, r, instance, v1.EventTypeWarning,
+				fmt.Sprintf("Failed to increment LastTriggerSyncID with TriggerSyncID: %d", instance.Spec.TriggerSyncID))
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+		msg := fmt.Sprintf("A full sync is already in progress. Ignoring this instance to trigger sync with triggersync ID: %d",
+			instance.Spec.TriggerSyncID)
+		log.Warn(msg)
+		recordEvent(ctx, r, instance, v1.EventTypeWarning, msg)
+		backOffDurationMapMutex.Lock()
+		delete(backOffDuration, instance.Name)
+		backOffDurationMapMutex.Unlock()
+		return reconcile.Result{}, nil
+	}
+
+	log.Infof("Reconciling trigger full sync with triggerSyncID: %d", instance.Spec.TriggerSyncID)
+	instance.Status.LastTriggerSyncID = instance.Spec.TriggerSyncID
+	instance.Status.InProgress = true
+	err = updateTriggerCsiFullSync(ctx, r.client, instance)
+	if err != nil {
+		recordEvent(ctx, r, instance, v1.EventTypeWarning,
+			fmt.Sprintf("Failed to update LastTriggerSyncID and Inprogress for TriggerSyncID: %d", instance.Spec.TriggerSyncID))
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+
+	startTime := time.Now()
+	triggerSyncID := instance.Spec.TriggerSyncID
+	var fullSyncErr error
+	if r.clusterFlavor == cnstypes.CnsClusterFlavorGuest {
+		fullSyncErr = syncer.PvcsiFullSync(ctx, syncer.MetadataSyncer)
+	} else {
+		fullSyncErr = syncer.CsiFullSync(ctx, syncer.MetadataSyncer)
+	}
+	err = r.client.Get(ctx, request.NamespacedName, instance)
```

Comment:
No need to call the get() again right? `updateTriggerCsiFullSync` is taking a pointer to struct of type `triggercsifullsyncv1alpha1.TriggerCsiFullSync`. You can use the same instance

Reply:
It is because we have updated the instance before with

```
instance.Status.LastTriggerSyncID = instance.Spec.TriggerSyncID
instance.Status.InProgress = true
```

before we set instance with success (or) failure, we need to get the latest instance. We want to get the instance right before the update so that no one we don't do updates on stale instances.

Reply:
update returns the latest instance
```
// Update updates the given obj in the Kubernetes cluster. obj must be a
// struct pointer so that obj can be updated with the content returned by the Server.
Update(ctx context.Context, obj runtime.Object, opts ...UpdateOption) error
```

Reply:
discussed offline. Update doesn't return latest instance.